File: Problemset/house-robber/house-robber.py

# @Title: 打家劫舍 (House Robber)
# @Author: cocofe
# @Date: 2021-04-15 23:56:42
# @Runtime: 32 ms
# @Memory: 14.9 MB

class Solution:
    def rob(self, nums: List[int]) -> int:
        pprev, prev = 0, 0
        for idx, num in enumerate(nums):
            curr = max(prev, pprev+num)
            pprev, prev = prev, curr
        return prev
        dp[i] = max(dp[i-2] + num, dp[i-3] + num)
            ans = max(ans, dp[i])
        return ans

File: Problemset/house-robber/house-robber.java

// @Title: 打家劫舍 (House Robber)
// @Author: cocofe
// @Date: 2021-07-24 18:37:16
// @Runtime: 0 ms
// @Memory: 35.8 MB

class Solution {
    public int rob(int[] nums) {
        int a = 0, b = 0, c = nums[0];
        for ( int i = 1; i< nums.length; i ++) {
            int tmp1 = a;
            int tmp2 = b;
            a = b;
            b = c;
            c = Math.max(tmp2, tmp1) + nums[i];
            // System.out.printf("i: %d  a:%d b:%d c:%d tmp1:%s tmp2:%s \n", i, a, b, c, tmp1, tmp2);
        }
        return Math.max(c, b);
    }
}
